File: scala/src/main/scala/atcoder/beginner/q117/ABC117D.scala
package atcoder.beginner.q117

object ABC117D {
  def main(args: Array[String]): Unit = {
    val sc = new java.util.Scanner(System.in)
    val N = sc.nextInt
    val K = sc.nextLong
    var A = List.empty[Long]
    (1 to N).foreach { f =>
      A = A :+ sc.nextLong
    }
    val abc117d = new ABC117D(N, K, A)
    println(abc117d.solve)
  }
}

class ABC117D(N: Int, K: Long, A: Seq[Long]) {

  def solve(): Long = {
    val N2 = N / 2
    var defaultBits = ~0L
    var xorBits = List.empty[Long]
    var bitsOn = List.empty[Boolean]
    var b = 1L
    while(K >= b) {
      xorBits = xorBits :+ b
      if ((K & b) > 0) bitsOn = bitsOn :+ true
      else bitsOn = bitsOn :+ false
      defaultBits = defaultBits ^ b;
      b = b << 1;
    }
    val bitLen = xorBits.length
    val bitsAggregate = Array.ofDim[Int](bitLen)

    var defaultSum = 0L
    A.foreach{ a =>
      defaultSum += defaultBits & a
      (0 to bitLen).foreach(i=> {
        if(((a >> i) & 1) == 1) {bitsAggregate(i) += 1}
      })
    }
    var sum1 = 0L
    var sum2 = 0L
    b = 1L
    bitsAggregate.zipWithIndex.foreach{
      case (x, i) => {
        val bitOff = x <= N2
        if (bitsOn(i) && bitOff) {sum1 += b * (N - bitsAggregate(i))}
        else {sum1 += b * bitsAggregate(i)}

        if (bitOff && i < bitLen - 1) {sum2 += b * (N - bitsAggregate(i))}
        else {sum2 += b * bitsAggregate(i)}
        b = b << 1
      }
    }

    if (sum1 > sum2) defaultSum + sum1
    else defaultSum + sum2
  }
}

File: java/src/main/java/atcoder/beginner/ABC117/ABC117D.java
package atcoder.beginner.ABC117;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class ABC117D {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int N = sc.nextInt();
    long K = sc.nextLong();
    long[] a = new long[N];
    for (int i = 0; i < N; i++) {
      a[i] = sc.nextLong();
    }
    ABC117D abc117D = new ABC117D();
    System.out.println(abc117D.solve(N, K, a));
  }

  public long solve(int N, long K, long[] a) {
    int N2 = N / 2;
    long defaultBits = ~0;
    List<Long> xorBits = new ArrayList();
    List<Boolean> bitsOn = new ArrayList();
    long b = 1;
    while (K >= b) {
      xorBits.add(b);
      if ((K & b) > 0) bitsOn.add(true);
      else bitsOn.add(false);
      defaultBits = defaultBits ^ b;
      b = b << 1;
    }
    int bitLen = xorBits.size();
    int[] bitsAggregate = new int[bitLen];


    long defaultSum = 0;
    for (long x : a) {
      defaultSum += defaultBits & x;
      for (int z = 0; z < xorBits.size(); z++) {
        if ((xorBits.get(z) & x) > 0) {
          bitsAggregate[z] += 1;
        }
      }
    }

    long sum1 = 0; // 一番左のbitが1
    long sum2 = 0; // 一番左のbitが0
    b = 1;
    for (int j = 0; j < bitsAggregate.length; j++) {
      boolean off = bitsAggregate[j] <= N2;
      if (bitsOn.get(j) && off) {
        sum1 += b * (N - bitsAggregate[j]);
      } else {
        sum1 += b * bitsAggregate[j];
      }
      if (off && j < bitsAggregate.length - 1) {
        sum2 += b * (N - bitsAggregate[j]);
      } else {
        sum2 += b * bitsAggregate[j];
      }
      b = b << 1;
    }
    if (sum1 > sum2 ) {
      return defaultSum+sum1;
    } else {
      return defaultSum+sum2;
    }
  }
}
